File: client/mia_companion/lib/screens/voice_call_screen.dart
import 'dart:async';
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';
import 'package:permission_handler/permission_handler.dart';

import '../services/api_service.dart';
import '../services/realtime_call_service.dart';
import '../theme/mia_theme.dart';
import '../widgets/mia_avatar.dart';

class VoiceCallScreen extends StatefulWidget {
  const VoiceCallScreen({super.key});

  @override
  State<VoiceCallScreen> createState() => _VoiceCallScreenState();
}

class _VoiceCallScreenState extends State<VoiceCallScreen>
    with SingleTickerProviderStateMixin {
  final _call = RealtimeCallService();
  Timer? _timer;
  int _seconds = 0;
  String _transcript = 'connecting…';
  List<int> _bars = List.filled(28, 10);
  bool _muted = false;
  bool _speaker = false;
  String? _error;
  bool _connected = false;
  bool _starting = true;

  late final AnimationController _pulse;

  @override
  void initState() {
    super.initState();
    _pulse = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 2200),
    )..repeat(reverse: true);
    _startCall();
  }

  Future<void> _startCall() async {
    final mic = await Permission.microphone.request();
    if (!mic.isGranted) {
      setState(() {
        _starting = false;
        _error = 'microphone permission is required for calls';
        _transcript = 'permission denied';
      });
      return;
    }

    _call.connectionController.stream.listen((state) {
      if (!mounted) return;
      if (state == CallConnectionState.ready) {
        setState(() {
          _connected = true;
          _starting = false;
          _error = null;
          _transcript = '…say something — i\'m listening';
        });
      } else if (state == CallConnectionState.error) {
        setState(() {
          _starting = false;
          _connected = false;
        });
      }
    });

    _call.transcriptController.stream.listen((t) {
      if (!mounted || t.isEmpty) return;
      setState(() {
        if (t.contains('error') ||
            t.contains('failed') ||
            t.contains('denied') ||
            t.contains('disconnect')) {
          _error = t;
        }
        _transcript = t.startsWith('…') ? t : '…$t';
      });
    });

    _call.levelController.stream.listen((level) {
      if (!mounted) return;
      setState(() {
        _bars = List.generate(28, (i) {
          final wobble = (math.sin(_seconds * 0.4 + i * 0.5) * 8).toInt();
          return (level + wobble).clamp(6, 56);
        });
      });
    });

    try {
      final session = await ApiService.instance.createRealtimeSession();
      final config = session['sessionConfig'] as Map<String, dynamic>;

      await _call.connect(
        wsUrl: session['wsUrl'] as String,
        token: session['token'] as String,
        sessionConfig: config,
      );

      _timer = Timer.periodic(const Duration(seconds: 1), (_) {
        if (mounted) setState(() => _seconds++);
      });
    } catch (e) {
      setState(() {
        _starting = false;
        _error = e.toString().replaceFirst('Exception: ', '');
        _transcript = 'couldn\'t connect';
      });
    }
  }

  String get _duration {
    final m = (_seconds ~/ 60).toString().padLeft(2, '0');
    final s = (_seconds % 60).toString().padLeft(2, '0');
    return '$m:$s';
  }

  Future<void> _endCall() async {
    await _call.hangUp();
    _timer?.cancel();
    if (mounted) Navigator.of(context).pop();
  }

  @override
  void dispose() {
    _timer?.cancel();
    _pulse.dispose();
    _call.hangUp();
    _call.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        decoration: const BoxDecoration(
          gradient: LinearGradient(
            begin: Alignment.topCenter,
            end: Alignment.bottomCenter,
            colors: [
              MiaColors.callGradientTop,
              MiaColors.callGradientMid,
              MiaColors.callGradientBottom,
            ],
            stops: [0.0, 0.45, 1.0],
          ),
        ),
        child: SafeArea(
          child: Column(
            children: [
              Padding(
                padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                child: Row(
                  children: [
                    _GlassButton(icon: Icons.arrow_back, onTap: _endCall),
                    Expanded(
                      child: Text(
                        'VOICE CALL',
                        textAlign: TextAlign.center,
                        style: GoogleFonts.inter(
                          color: Colors.white54,
                          fontSize: 11,
                          letterSpacing: 2,
                          fontWeight: FontWeight.w600,
                        ),
                      ),
                    ),
                    const SizedBox(width: 44),
                  ],
                ),
              ),
              const SizedBox(height: 32),
              AnimatedBuilder(
                animation: _pulse,
                builder: (context, child) {
                  final scale = 1.0 + (_pulse.value * 0.06);
                  return Transform.scale(scale: scale, child: child);
                },
                child: Container(
                  width: 168,
                  height: 168,
                  decoration: BoxDecoration(
                    shape: BoxShape.circle,
                    border: Border.all(
                      color: Colors.white.withValues(alpha: 0.35),
                      width: 2,
                    ),
                    boxShadow: [
                      BoxShadow(
                        color: MiaColors.accent.withValues(alpha: 0.45),
                        blurRadius: 48,
                        spreadRadius: 4,
                      ),
                    ],
                  ),
                  child: Stack(
                    fit: StackFit.expand,
                    children: [
                      const MiaAvatar(size: 168),
                      if (_starting)
                        Container(
                          decoration: BoxDecoration(
                            shape: BoxShape.circle,
                            color: Colors.black.withValues(alpha: 0.45),
                          ),
                          child: const Center(
                            child: CircularProgressIndicator(
                              color: Colors.white70,
                              strokeWidth: 2,
                            ),
                          ),
                        )
                      else if (!_connected)
                        Container(
                          decoration: BoxDecoration(
                            shape: BoxShape.circle,
                            color: Colors.black.withValues(alpha: 0.35),
                          ),
                          child: Icon(
                            Icons.signal_wifi_off_rounded,
                            size: 40,
                            color: Colors.white.withValues(alpha: 0.85),
                          ),
                        ),
                    ],
                  ),
                ),
              ),
              const SizedBox(height: 28),
              Text(
                'Mia',
                style: GoogleFonts.playfairDisplay(
                  fontSize: 38,
                  color: Colors.white,
                  fontWeight: FontWeight.w600,
                ),
              ),
              const SizedBox(height: 12),
              Row(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  Container(
                    width: 8,
                    height: 8,
                    decoration: BoxDecoration(
                      color: _connected ? MiaColors.online : Colors.orange,
                      shape: BoxShape.circle,
                    ),
                  ),
                  const SizedBox(width: 8),
                  Text(
                    _connected
                        ? 'on the line · $_duration'
                        : _starting
                            ? 'connecting…'
                            : 'not connected',
                    style: GoogleFonts.inter(
                      color: Colors.white70,
                      fontSize: 13,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 32),
              SizedBox(
                height: 52,
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  crossAxisAlignment: CrossAxisAlignment.end,
                  children: _bars.map((h) {
                    return AnimatedContainer(
                      duration: const Duration(milliseconds: 100),
                      width: 3,
                      height: h.toDouble(),
                      margin: const EdgeInsets.symmetric(horizontal: 2),
                      decoration: BoxDecoration(
                        color: Colors.white.withValues(alpha: 0.9),
                        borderRadius: BorderRadius.circular(2),
                      ),
                    );
                  }).toList(),
                ),
              ),
              const SizedBox(height: 28),
              Padding(
                padding: const EdgeInsets.symmetric(horizontal: 24),
                child: Container(
                  width: double.infinity,
                  constraints: const BoxConstraints(minHeight: 72),
                  padding: const EdgeInsets.symmetric(
                    horizontal: 20,
                    vertical: 18,
                  ),
                  decoration: BoxDecoration(
                    color: Colors.black.withValues(alpha: 0.28),
                    borderRadius: BorderRadius.circular(22),
                    border: Border.all(
                      color: Colors.white.withValues(alpha: 0.08),
                    ),
                  ),
                  child: Text(
                    _error ?? _transcript,
                    style: GoogleFonts.inter(
                      color: Colors.white.withValues(alpha: 0.95),
                      fontSize: 15,
                      fontStyle: FontStyle.italic,
                      height: 1.45,
                    ),
                    textAlign: TextAlign.center,
                  ),
                ),
              ),
              const Spacer(),
              Padding(
                padding: const EdgeInsets.fromLTRB(32, 0, 32, 28),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: [
                    _CallAction(
                      icon: _muted ? Icons.mic_off_rounded : Icons.mic_rounded,
                      label: 'MUTE',
                      active: _muted,
                      onTap: () async {
                        final next = !_muted;
                        setState(() => _muted = next);
                        await _call.setMuted(next);
                      },
                    ),
                    _CallAction(
                      icon: _speaker
                          ? Icons.volume_up_rounded
                          : Icons.hearing_rounded,
                      label: _speaker ? 'SPEAKER' : 'EARPIECE',
                      active: _speaker,
                      onTap: () async {
                        final next = !_speaker;
                        setState(() => _speaker = next);
                        await _call.setSpeaker(next);
                      },
                    ),
                    GestureDetector(
                      onTap: _endCall,
                      child: Container(
                        width: 68,
                        height: 68,
                        decoration: BoxDecoration(
                          shape: BoxShape.circle,
                          color: const Color(0xFFE53935),
                          boxShadow: [
                            BoxShadow(
                              color: const Color(0xFFE53935)
                                  .withValues(alpha: 0.55),
                              blurRadius: 20,
                              spreadRadius: 2,
                            ),
                          ],
                        ),
                        child: const Icon(
                          Icons.call_end_rounded,
                          color: Colors.white,
                          size: 30,
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _GlassButton extends StatelessWidget {
  const _GlassButton({required this.icon, required this.onTap});

  final IconData icon;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return Material(
      color: Colors.white.withValues(alpha: 0.14),
      shape: const CircleBorder(),
      child: InkWell(
        onTap: onTap,
        customBorder: const CircleBorder(),
        child: SizedBox(
          width: 44,
          height: 44,
          child: Icon(icon, color: Colors.white, size: 22),
        ),
      ),
    );
  }
}

class _CallAction extends StatelessWidget {
  const _CallAction({
    required this.icon,
    required this.label,
    required this.onTap,
    this.active = false,
  });

  final IconData icon;
  final String label;
  final VoidCallback onTap;
  final bool active;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onTap,
      child: Column(
        children: [
          Container(
            width: 54,
            height: 54,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: active
                  ? Colors.white.withValues(alpha: 0.22)
                  : Colors.white.withValues(alpha: 0.1),
            ),
            child: Icon(icon, color: Colors.white.withValues(alpha: 0.85)),
          ),
          const SizedBox(height: 8),
          Text(
            label,
            style: GoogleFonts.inter(
              color: Colors.white54,
              fontSize: 10,
              letterSpacing: 1.2,
              fontWeight: FontWeight.w600,
            ),
          ),
        ],
      ),
    );
  }
}

File: client/mia_android/app/src/main/kotlin/com/mia/companion/ui/screens/VoiceCallScreen.kt
package com.mia.companion.ui.screens

import android.Manifest
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.CallEnd
import androidx.compose.material.icons.outlined.Hearing
import androidx.compose.material.icons.outlined.Mic
import androidx.compose.material.icons.outlined.MicOff
import androidx.compose.material.icons.outlined.SignalWifiOff
import androidx.compose.material.icons.outlined.VolumeUp
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.mia.companion.data.ApiClient
import com.mia.companion.ui.components.MiaAvatar
import com.mia.companion.ui.theme.MiaColors
import com.mia.companion.ui.theme.MiaTypography
import com.mia.companion.voice.CallConnectionState
import com.mia.companion.voice.RealtimeVoiceEngine
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.sin

@Composable
fun VoiceCallScreen(onEnd: () -> Unit) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val engine = remember { RealtimeVoiceEngine(context) }

    var seconds by remember { mutableIntStateOf(0) }
    var transcript by remember { mutableStateOf("connecting…") }
    var error by remember { mutableStateOf<String?>(null) }
    var starting by remember { mutableStateOf(true) }
    var connected by remember { mutableStateOf(false) }
    var muted by remember { mutableStateOf(false) }
    var speaker by remember { mutableStateOf(false) }
    var bars by remember { mutableStateOf(List(28) { 10 }) }

    val connection by engine.connection.collectAsState()

    val micPermission = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission(),
    ) { granted ->
        if (!granted) {
            starting = false
            error = "microphone permission is required for calls"
            transcript = "permission denied"
        } else {
            scope.launch {
                connectCall(engine) { starting = it }
            }
        }
    }

    LaunchedEffect(Unit) { micPermission.launch(Manifest.permission.RECORD_AUDIO) }

    LaunchedEffect(Unit) {
        while (true) {
            delay(1000)
            seconds++
        }
    }

    LaunchedEffect(Unit) {
        engine.transcript.collect { t ->
            if (t.contains("error", true) || t.contains("failed", true) ||
                t.contains("denied", true) || t.contains("disconnect", true)
            ) error = t
            transcript = if (t.startsWith("…")) t else "…$t"
        }
    }

    LaunchedEffect(Unit) {
        engine.level.collect { level ->
            bars = List(28) { i ->
                val wobble = (sin(seconds * 0.4 + i * 0.5) * 8).toInt()
                (level + wobble).coerceIn(6, 56)
            }
        }
    }

    LaunchedEffect(connection) {
        when (connection) {
            CallConnectionState.READY -> {
                starting = false
                connected = true
                error = null
                transcript = "…say something — i'm listening"
            }
            CallConnectionState.ERROR -> {
                starting = false
                connected = false
            }
            else -> Unit
        }
    }

    DisposableEffect(Unit) {
        onDispose {
            scope.launch { engine.hangUp() }
            engine.release()
        }
    }

    val pulse by rememberInfiniteTransition(label = "pulse").animateFloat(
        0f, 1f, infiniteRepeatable(tween(2200), RepeatMode.Reverse), label = "p",
    )
    val scale = 1f + pulse * 0.06f

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(
                Brush.verticalGradient(
                    0f to MiaColors.CallGradientTop,
                    0.45f to MiaColors.CallGradientMid,
                    1f to MiaColors.CallGradientBottom,
                ),
            ),
    ) {
        Column(Modifier.fillMaxSize()) {
            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 8.dp, vertical = 4.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                GlassButton(Icons.AutoMirrored.Filled.ArrowBack) {
                    scope.launch { engine.hangUp(); onEnd() }
                }
                Text(
                    "VOICE CALL",
                    style = MiaTypography.inter(11f, FontWeight.SemiBold, Color.White.copy(0.54f), letterSpacing = 2f),
                    modifier = Modifier.weight(1f),
                    textAlign = TextAlign.Center,
                )
                Spacer(Modifier.width(44.dp))
            }
            Spacer(Modifier.height(32.dp))
            Box(
                modifier = Modifier
                    .size(168.dp)
                    .scale(scale)
                    .align(Alignment.CenterHorizontally)
                    .border(2.dp, Color.White.copy(alpha = 0.35f), CircleShape)
                    .shadow(48.dp, CircleShape, spotColor = MiaColors.Accent.copy(alpha = 0.45f)),
                contentAlignment = Alignment.Center,
            ) {
                MiaAvatar(size = 168.dp)
                if (starting) {
                    Box(
                        Modifier
                            .fillMaxSize()
                            .clip(CircleShape)
                            .background(Color.Black.copy(0.45f)),
                        contentAlignment = Alignment.Center,
                    ) {
                        CircularProgressIndicator(color = Color.White.copy(0.7f), strokeWidth = 2.dp)
                    }
                } else if (!connected) {
                    Box(
                        Modifier
                            .fillMaxSize()
                            .clip(CircleShape)
                            .background(Color.Black.copy(0.35f)),
                        contentAlignment = Alignment.Center,
                    ) {
                        Icon(Icons.Outlined.SignalWifiOff, null, tint = Color.White.copy(0.85f), modifier = Modifier.size(40.dp))
                    }
                }
            }
            Spacer(Modifier.height(28.dp))
            Text("Mia", style = MiaTypography.serifTitle(38f).copy(color = Color.White), modifier = Modifier.align(Alignment.CenterHorizontally))
            Spacer(Modifier.height(12.dp))
            Row(
                Modifier.align(Alignment.CenterHorizontally),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(
                    Modifier
                        .size(8.dp)
                        .clip(CircleShape)
                        .background(if (connected) MiaColors.Online else Color(0xFFFF9800)),
                )
                Spacer(Modifier.width(8.dp))
                Text(
                    when {
                        connected -> "on the line · ${formatDuration(seconds)}"
                        starting -> "connecting…"
                        else -> "not connected"
                    },
                    style = MiaTypography.inter(13f, color = Color.White.copy(0.7f)),
                )
            }
            Spacer(Modifier.height(32.dp))
            Row(
                Modifier
                    .height(52.dp)
                    .align(Alignment.CenterHorizontally),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.Bottom,
            ) {
                bars.forEach { h ->
                    Box(
                        Modifier
                            .padding(horizontal = 2.dp)
                            .width(3.dp)
                            .height(h.dp)
                            .background(Color.White.copy(0.9f), RoundedCornerShape(2.dp)),
                    )
                }
            }
            Spacer(Modifier.height(28.dp))
            Text(
                error ?: transcript,
                style = MiaTypography.inter(15f, color = Color.White.copy(0.95f), lineHeight = 21.75f)
                    .copy(fontStyle = FontStyle.Italic),
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .padding(horizontal = 24.dp)
                    .fillMaxWidth()
                    .background(Color.Black.copy(0.28f), RoundedCornerShape(22.dp))
                    .border(1.dp, Color.White.copy(0.08f), RoundedCornerShape(22.dp))
                    .padding(horizontal = 20.dp, vertical = 18.dp),
            )
            Spacer(Modifier.weight(1f))
            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 32.dp, vertical = 28.dp),
                horizontalArrangement = Arrangement.SpaceEvenly,
                verticalAlignment = Alignment.Top,
            ) {
                CallAction(
                    if (muted) Icons.Outlined.MicOff else Icons.Outlined.Mic,
                    "MUTE",
                    muted,
                ) { muted = !muted; engine.setMuted(muted) }
                Box(
                    Modifier
                        .size(68.dp)
                        .shadow(20.dp, CircleShape, spotColor = Color(0xFFE53935).copy(0.55f))
                        .clip(CircleShape)
                        .background(Color(0xFFE53935))
                        .clickable {
                            scope.launch { engine.hangUp(); onEnd() }
                        },
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(Icons.Filled.CallEnd, null, tint = Color.White, modifier = Modifier.size(30.dp))
                }
                CallAction(
                    if (speaker) Icons.Outlined.VolumeUp else Icons.Outlined.Hearing,
                    if (speaker) "SPEAKER" else "EARPIECE",
                    speaker,
                ) { speaker = !speaker; engine.setSpeaker(speaker) }
            }
        }
    }
}

@Composable
private fun GlassButton(icon: androidx.compose.ui.graphics.vector.ImageVector, onClick: () -> Unit) {
    Box(
        Modifier
            .size(44.dp)
            .clip(CircleShape)
            .background(Color.White.copy(0.14f))
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Icon(icon, null, tint = Color.White, modifier = Modifier.size(22.dp))
    }
}

@Composable
private fun CallAction(
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    label: String,
    active: Boolean,
    onClick: () -> Unit,
) {
    Column(horizontalAlignment = Alignment.CenterHorizontally, modifier = Modifier.clickable(onClick = onClick)) {
        Box(
            Modifier
                .size(54.dp)
                .clip(CircleShape)
                .background(if (active) Color.White.copy(0.22f) else Color.White.copy(0.1f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(icon, null, tint = Color.White.copy(0.85f))
        }
        Spacer(Modifier.height(8.dp))
        Text(label, style = MiaTypography.inter(10f, FontWeight.SemiBold, Color.White.copy(0.54f), letterSpacing = 1.2f))
    }
}

private suspend fun connectCall(engine: RealtimeVoiceEngine, onStarting: (Boolean) -> Unit) {
    try {
        val session = ApiClient.createRealtimeSession()
        engine.connect(
            wsUrl = session.wsUrl,
            token = session.token,
            sessionConfig = session.sessionConfig ?: emptyMap(),
            sessionPreconfigured = session.sessionPreconfigured,
        )
        onStarting(false)
    } catch (_: Exception) {
        onStarting(false)
    }
}

private fun formatDuration(seconds: Int): String {
    val m = (seconds / 60).toString().padStart(2, '0')
    val s = (seconds % 60).toString().padStart(2, '0')
    return "$m:$s"
}
